allennlp/experiments/drivers/train_driver.py

```diff
@@ -0,0 +1,182 @@
+                    loss = output_dict["loss"]
+                    loss.backward()
+                    train_loss += loss.data.numpy()
+                except KeyError:
+                    raise ConfigurationError("The model you are trying to optimize does not contain a"
+                                             " 'loss' key in the output of model.forward(inputs).")
+                # TODO(Mark): Do we care about optimizers which require a closure?
+                self._optimizer.step()
+
+            # Switch to evaluation mode.
+            if self._validation_dataset is not None:
+                self._model.eval()
+                val_generator = self._iterator(self._validation_dataset, num_epochs=1)
+                for batch in tqdm.tqdm(val_generator):
+                    tensor_batch = arrays_to_variables(batch, self._cuda_device)
+                    val_output_dict = self._model.forward(tensor_batch)
+                    loss = val_output_dict["loss"]
+                    val_loss += loss.data.numpy()
+
+            # TODO(Mark): Add user specified metrics here, maybe a "metrics" key?
+            logger.info("Training Loss: %3f    Validation Loss: %3f ", train_loss, val_loss)
+            if self._serialization_prefix:
+                self._save_checkpoint(epoch)
+
+    def _save_checkpoint(self,
+                         epoch: int,
+                         is_best: Optional[bool] = None) -> None:
+        model_path = os.path.join(self._serialization_prefix, "model_state_epoch_{}.th".format(epoch))
+        model_state = self._model.state_dict()
+        torch.save(model_state, model_path)
+
+        training_state = {'epoch': epoch, 'optimizer': self._optimizer.state_dict()}
+        torch.save(training_state, os.path.join(self._serialization_prefix,
+                                                "training_state_epoch_{}.th".format(epoch)))
+        if is_best:
+            shutil.copy(model_path, os.path.join(model_path, "best.th"))
+
+    def _restore_checkpoint(self) -> int:
+
+        serialization_files = os.listdir(self._serialization_prefix)
+        model_checkpoints = [x for x in serialization_files if "model_state_epoch" in x]
+        epoch_to_load = max([int(x.split("model_state_epoch_")[-1].strip(".th")) for x in model_checkpoints])
+
+        model_state = torch.load(os.path.join(self._serialization_prefix,
+                                              "model_state_epoch_{}.th".format(epoch_to_load)))
+        training_state = torch.load(os.path.join(self._serialization_prefix,
+                                                 "training_state_epoch_{}.th".format(epoch_to_load)))
+        self._model.load_state_dict(model_state)
+        self._optimizer.load_state_dict(training_state["optimizer"])
+        return training_state["epoch"]
+
+    @classmethod
+    def from_params(cls, params: Params) -> 'TrainDriver':
+        dataset_reader = DatasetReader.from_params(params.pop('dataset_reader'))
```

Comment:
I think this whole method should go in the `run.py` script.  The stuff done here isn't training, it's setting up training.  And then there are two entry points to using our trainer: (1) you use our `run.py` script, which takes a parameter file as an argument, or (2) you construct the objects yourself and call `Trainer.train()`.